# SQL

Привет! Меня зовут Анастасия!

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1">Описание проекта</a></span></li><li><span><a href="#Задания" data-toc-modified-id="Задания-2">Задания</a></span></li><li><span><a href="#Как-выполнить-задание?" data-toc-modified-id="Как-выполнить-задание?-3">Как выполнить задание?</a></span></li><li><span><a href="#Исследование-таблиц" data-toc-modified-id="Исследование-таблиц-4">Исследование таблиц</a></span></li><li><span><a href="#1.-Cколько-книг-вышло-после-1-января-2000-года;" data-toc-modified-id="1.-Cколько-книг-вышло-после-1-января-2000-года;-5">1. Cколько книг вышло после 1 января 2000 года;</a></span></li><li><span><a href="#2.-Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;" data-toc-modified-id="2.-Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;-6">2. Для каждой книги посчитайте количество обзоров и среднюю оценку;</a></span></li><li><span><a href="#3.-Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц;" data-toc-modified-id="3.-Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц;-7">3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц;</a></span></li><li><span><a href="#4.-Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;" data-toc-modified-id="4.-Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;-8">4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;</a></span></li><li><span><a href="#5.-Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок." data-toc-modified-id="5.-Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок.-9">5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</a></span></li></ul></div>

### Описание проекта

**Цель:** проанализировать базу данных сервиса для чтения книг для формулирования ценностного предложения для нового продукта.

**Задания:**

**1)** Посчитайте, сколько книг вышло после 1 января 2000 года;

**2)** Для каждой книги посчитайте количество обзоров и среднюю оценку;

**3)** Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

**4)** Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

**5)** Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.


**Описание данных:**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [7]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [16]:
# запишем функцию для запросов: 

def select(query):
    con=engine.connect()
    return pd.io.sql.read_sql(sql=text(query), con = con)

### Исследование таблиц

In [21]:
query = '''SELECT * FROM books LIMIT 5'''
select(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [26]:
query = '''SELECT COUNT (*) FROM books'''
select(query)

count
0   1000

**В таблице с книгами представлена информация о 1 000 книг.**

In [22]:
query = '''SELECT * FROM authors LIMIT 5'''
select(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [24]:
query = '''SELECT * FROM publishers LIMIT 5'''
select(query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [23]:
query = '''SELECT * FROM ratings LIMIT 5'''
select(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [25]:
query = '''SELECT * FROM reviews LIMIT 5'''
select(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### 1. Cколько книг вышло после 1 января 2000 года;

In [20]:
query = '''SELECT COUNT (*) AS Количество_книг FROM books 
        WHERE publication_date > date('2000-01-01')'''

select(query)

Количество_книг
0              819

**Из 1 000 книг 80% (819 штук) вышли после 1 января 2000 года.**

### 2. Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [29]:
query = ''' WITH table_1 AS (
            SELECT * FROM 
            (SELECT book_id, round(AVG(rating), 2) AS Средняя_оценка FROM ratings
             GROUP BY book_id) AS t_1 FULL JOIN (SELECT book_id, COUNT(*) AS Количество_обзоров FROM ratings
                                                  GROUP BY book_id) AS t_2 USING (book_id))
            SELECT author, title, Средняя_оценка, Количество_обзоров
            FROM books LEFT JOIN table_1 USING (book_id)
                       LEFT JOIN authors USING (author_id)
            ORDER BY Средняя_оценка DESC'''

select(query)

author  \
0                                      James Patterson   
1                                       Steven Johnson   
2                                        Tamora Pierce   
3                                        James Clavell   
4                                       Nigella Lawson   
..                                                 ...   
995                                 Thomas L. Friedman   
996  William S. Burroughs/Oliver  Harris/Allen Gins...   
997                                    Joseph J. Ellis   
998                                  Christina Schwarz   
999                                       Jodi Picoult   

                                                 title  Средняя_оценка  \
0                 Pop Goes the Weasel (Alex Cross  #5)            5.00   
1    The Ghost Map: The Story of London's Most Terr...            5.00   
2    In the Hand of the Goddess (Song of the Liones...            5.00   
3                             Tai-Pan (Asian Saga  #2)            5.00   
4    How to Be a Domestic Goddess: Baking and the A...            5.00   
..                                                 ...             ...   
995  The World Is Flat: A Brief History of the Twen...            2.25   
996                                              Junky            2.00   
997                  His Excellency: George Washington            2.00   
998                                      Drowning Ruth            2.00   
999                               Harvesting the Heart            1.50   

     Количество_обзоров  
0                     2  
1                     2  
2                     3  
3                     2  
4                     2  
..                  ...  
995                   4  
996                   2  
997                   2  
998                   3  
999                   2  

[1000 rows x 4 columns]

**Средняя оценка книг в диапазоне от 1.5 до 5 баллов.** Самый низкий рейтинг у книги Harvesting the Heart by Jodi Picoult. 

### 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц;

In [32]:
query = ''' SELECT publisher, Количество_книг
            FROM (
                  SELECT publisher_id, COUNT (*) AS Количество_книг 
                  FROM  books
                  WHERE num_pages > 50
                  GROUP BY publisher_id
                  ORDER BY Количество_книг DESC
                  LIMIT 1) AS t_1 
                              INNER JOIN publishers USING (publisher_id)'''

select(query)

publisher  Количество_книг
0  Penguin Books               42

Больше всего книг выпустило издание - **Penguin Books**.

### 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [46]:
query = '''WITH table_1 AS (
                 SELECT book_id, COUNT(rating) AS Количество_оценок, round(AVG(rating), 2) AS Оценки
                 FROM ratings
                 GROUP BY book_id
                 HAVING COUNT(rating) >= 50) 
           SELECT author, AVG(Оценки) AS Средняя_оценка
           FROM table_1 INNER JOIN books USING (book_id)
                        INNER JOIN authors USING (author_id)
           GROUP BY author
           ORDER BY Средняя_оценка DESC
           LIMIT 1'''

select(query)

author  Средняя_оценка
0  J.K. Rowling/Mary GrandPré           4.285

**Автор с самой высокой средней оценкой книг - J.K. Rowling.**

### 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [50]:
query = '''WITH table_1 AS (
           SELECT username, COUNT(review_id) AS Количество_обзоров
           FROM (
                SELECT username, COUNT(rating) AS Количество_оценок FROM ratings
                GROUP BY username
                HAVING COUNT(rating) > 48) AS t_1 
                                                INNER JOIN reviews USING (username)
           GROUP BY username)
           SELECT AVG(Количество_обзоров) AS Средн_количество_обзоров
           FROM table_1'''


select(query)

Средн_количество_обзоров
0                      24.0

Пользователи, которые поставили более 48 оценок, имеют в среднем 24 обзора.